Spring-data

ppapou opened the pull request:
- the example of spring boot data application has been added

Commit:
Spring-data

 - the example of spring boot data application has been added

oilin-clgx commented on data/src/main/java/com/example/data/CustomerRepository.java:
```diff
@@ -0,0 +1,12 @@
+
+    List<Customer> findByLastName(String lastName);
+
+    Customer findById(long id);
```
I believe it is redundant. It is already a part of  CrudRepository and you do not need to redefine it if you do not want to change it somehow

ppapou replied:
the repository method returns the Optional object. 
For demo version the Customer obj is returned

oleksandr-ilin replied:
Question: what happens if there is no object with such Id?

oilin-clgx commented on data/src/main/java/com/example/data/DataApplication.java:
```diff
@@ -0,0 +1,53 @@
+		SpringApplication.run(DataApplication.class);
+	}
+
+	@Bean
```
Usually that is bad idea to put some extra code into the Application. Keep it simple. Instead feel free to define separate `@Component` with appropriated interface to run console apps

ppapou replied:
it is just a demo, the service component will be created

oilin-clgx reviewed (changes requested):
Some comments was added + I can't find `gradle` scripts. How it is supposed to be built?

Another good idea you might to consider could be to use `Lombok` to reduce the boilerplate.

Commit:
Spring data

 - project strucuture has adusted
 - gradle files have been uploaded

Commit:
Gradle propeties has added

Commit:
Project structure

 - data packges names have been overwritten
 - service layer has adde

Commit:
Minor note has added

Commit:
Spring boot data app

 - controller, customer service are added
 - hash, equals methods have been overwritten for Customer class

Commit:
Application title

oleksandr-ilin commented on data/build.gradle:
```diff
@@ -0,0 +1,28 @@
+
+dependencies {
+	implementation 'org.springframework.boot:spring-boot-starter-data-jdbc'
+	testImplementation 'org.springframework.boot:spring-boot-starter-test'
```
I would recommend organize `dependencies` section in some way. E.g. annotationProcessor and compileOnly go first, then all `implementation` and `runtimeOnly' and all `testImplementation` go last.

Another good idea is to separate all that logical parts by extra empty line

oleksandr-ilin commented on data/src/main/java/com/example/data/CustomerDataApplication.java:
```diff
@@ -0,0 +1,51 @@
+
+	//TODO: the demo method will be removed
+	@Bean
+	public CommandLineRunner demo(CustomerRepository repository) {
```
Please move the bean into the separate own file or some Configuration bean.
Keep Application bean simple (just main function)
All other stuff like bean definitions, functional and feature annotations, console runners etc keep in the appropriated files (separate or inside Configuration bean)

oleksandr-ilin commented on data/src/main/java/com/example/data/backendapi/controller/CustomerController.java:
```diff
@@ -0,0 +1,22 @@
+@RestController
+public class CustomerController {
+
+    @Autowired
```
This is not recommended way to define injection. Prefer constructor injection instead.
One of the good reason for that it could simplify the unit testing. Also that will help you to keep the class under control and do not bloat it with many dependencies violating Low Coupling, Single Responsibility, Interface segregation principles.

oleksandr-ilin commented on data/src/main/java/com/example/data/backendapi/dto/Customer.java:
```diff
@@ -0,0 +1,63 @@
+/**
+ * CLient's objects with personal information
+ */
+@Entity
```
I do not think it is DTO object. That is fair Entity so prefer to move it into the appropriated package

oleksandr-ilin commented on data/src/main/java/com/example/data/backendapi/dto/Customer.java:
```diff
@@ -0,0 +1,63 @@
+
+    @Id
+    @GeneratedValue(strategy=GenerationType.AUTO)
+    private Long id;
```
Consider to use `Lombok` annotation. That might significantly reduce boilerplate and make it easy to maintain. For caparison it could be just like that one:
```
@Entity
@Value
@NoArgsConstructor
@AllArgsConstructor
public class Customer {

    @Id
    @GeneratedValue(strategy=GenerationType.AUTO)
    Long id;
    String firstName;
    String lastName;
}
```

This is Data Object stereotype. So it must not contain the logic. Using Lombok here considered as an acceptable tradeoff.

Also you can take a look on new Java feature: `records`

oleksandr-ilin commented on data/src/main/java/com/example/data/backendapi/service/CustomerService.java:
```diff
@@ -0,0 +1,36 @@
+
+import java.util.Optional;
+
+@Scope("singleton")
```
Prefer @Service or @Component instead. They are all singletons by default anyway

oleksandr-ilin commented on data/src/main/java/com/example/data/backendapi/service/CustomerService.java:
```diff
@@ -0,0 +1,36 @@
+public class CustomerService {
+
+    @Autowired
+    private CustomerRepository repository;
```
Make it final. And with Lombok you just can add annotation `@RequiredArgsConstructor` so it would maintain all your final fields as an arguments for injection in the constructor. Easy to maintain!

oleksandr-ilin commented on data/src/main/java/com/example/data/backendapi/controller/CustomerController.java:
```diff
@@ -0,0 +1,22 @@
+    Customer one(@PathVariable Long id) {
+
+        return customerService.findById(id)
+                .orElseThrow(() -> new CustomerNotFoundException(id));
```
If you like it might be a good practice to add some exertion handler for `CustomerNotFoundException` so the user get sone nice output instead of default spring one.

Also consider the alternative and define your error handler right in the controller. Make sure you clearly understand your options in terms of handling errors.

oleksandr-ilin commented on data/src/test/java/com/example/data/DataApplicationTests.java:
```diff
@@ -0,0 +1,13 @@
+import org.springframework.boot.test.context.SpringBootTest;
+
+@SpringBootTest
+class DataApplicationTests {
```
I would recommend to move away into separate sourceSet all heavy integration tests. 
Try to create a new sourceSet and task `integration-test` and also try to add the integration test to test out your controller. That would be useful.

Here should be unit tests only. 
The unit test should run really fast and test only tested unit. Loading the Spring Context and loading many other units breaks that concept.

Your unit test must import just the tested class and all the dependencies must be just mocked.
Check it out what is `test pyramid`.

oleksandr-ilin reviewed (changes requested):
Looks very good. But I have added some change requests which I believe would be useful for you.

Commit:
Spring-boot data

 - load repository runner has moved to separated package
 - contorller method are added

Commit:
Application import adjusted

Commit:
Spring-boot data

 - RESTFUL controller has been improved
